### PR TITLE
Make generic TypeAliasType specifications produce different schema definitions

### LINKED ISSUE #7840: PEP 695 generic type alias does not work
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

PEP 695 generic type alias does not seem to work. JSON schema also does not look right. Looks like the first type definition is somehow cached and re-used incorrectly. See MRE, python 3.12+. I also tried v2.4.2, but the output is the same

### Example Code

```Python
from collections.abc import Sequence
from typing import Annotated

from pydantic import BaseModel, Field, ValidationError

type NonEmptySequence[T] = Annotated[
    Sequence[T],
    Field(min_length=1, max_length=1024)
]


class ModelA(BaseModel):
    a: int


class ModelB(BaseModel):
    b: str


class Model(BaseModel):
    a_s: NonEmptySequence[ModelA]
    b_s: NonEmptySequence[ModelB]


print(Model.model_json_schema())

try:
    model = Model(
        a_s=[ModelA(a=1)],
        b_s=[ModelB(b='test')]
    )
except ValidationError as error:
    print(error)
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.3.0
        pydantic-core version: 2.6.3
          pydantic-core build: profile=release pgo=false
                 install path: /.venv/lib/python3.12/site-packages/pydantic
               python version: 3.12.0 (v3.12.0:0fb18b02c8, Oct  2 2023, 09:45:56) [Clang 13.0.0 (clang-1300.0.29.30)]
                     platform: macOS-13.0.1-arm64-arm-64bit
     optional deps. installed: ['email-validator', 'typing-extensions']
```

### PR DESCRIPTION
**Pull request by alexdrydew**
## Change Summary

This PR fixes #7840: now different specifications of a generic TypeAliasType produce different definitions in pydantic schema which prevents "overwrites" of other specifications with the same origin.

## Related issue number

* Fix #7840 

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [x] Unit tests for the changes exist
* [x] Tests pass on CI
* [x] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Selected Reviewer: @sydney-runkle

### CODE REVIEW

**Comment by alexdrydew** [comment]
please review

**Review by sydney-runkle — CHANGES_REQUESTED** [review]
LGTM overall, left a few nit picky change requests.

Thanks for the contribution ⭐!!

File: pydantic/_internal/_typing_extra.py [hunk 1]
```diff
@@ -433,3 +433,10 @@ def is_dataclass(_cls: type[Any]) -> TypeGuard[type[StandardDataclass]]:
 
 def origin_is_type_alias_type(origin: Any) -> TypeGuard[TypeAliasType]:
     return isinstance(origin, TypeAliasType)
+
+
+def is_generic_alias(type_: type[Any]) -> bool:
+    if sys.version_info >= (3, 10):
+        return isinstance(type_, (types.GenericAlias, typing._GenericAlias))  # type: ignore[attr-defined]
+    else:
+        return isinstance(type_, typing._GenericAlias)  # type: ignore
```
Comment (dmontagu) [inline pydantic/_internal/_typing_extra.py:442]:
```suggestion
if sys.version_info >= (3, 10):
    def is_generic_alias(type_: type[Any]) -> bool:
        return isinstance(type_, (types.GenericAlias, typing._GenericAlias))  # type: ignore[attr-defined] 
else:
    def is_generic_alias(type_: type[Any]) -> bool:
        return isinstance(type_, typing._GenericAlias)  # type: ignore
```
Super minor, but eliminates a runtime branch 🤷‍♂️

File: tests/test_type_alias_type.py [hunk 2]
```diff
@@ -185,14 +185,15 @@ def test_type_alias_annotated_defs() -> None:
...
+            {'$ref': '#/$defs/MyList_int__MaxLen_max_length_1_'},
+            {'$ref': '#/$defs/MyList_int__MaxLen_max_length_1_'},
         ],
         'maxItems': 2,
-        '$defs': {'MyList_MaxLen_max_length_1_': {'type': 'array', 'items': {'type': 'integer'}, 'maxItems': 1}},
+        '$defs': {'MyList_int__MaxLen_max_length_1_': {'type': 'array', 'items': {'type': 'integer'}, 'maxItems': 1}},
     }
 
 
+@pytest.mark.xfail(sys.version_info >= (3, 12), reason="TypeAliasType doesn't have __qualname__ yet")
```
Comment (dmontagu) [inline tests/test_type_alias_type.py:196]: It fails for python _newer_ than (or equal to) 3.12? Saying "doesn't have qualname yet" makes me think the failure would happen for older-than-3.12, not 3.12 or newer.

Could you explain what's going wrong in a bit more detail?

Comment (dmontagu) [reply tests/test_type_alias_type.py:196]: Nevermind I see this was just copying what existing tests were doing

Comment (dmontagu) [reply tests/test_type_alias_type.py:196]: I see a way to resolve the need for these xfails, but I think that should be done in a separate PR, so ignore this comment.

Comment (sydney-runkle) [inline tests/test_type_alias_type.py:196]: Based on the changes in #7929, I think you can remove this xfail marker

**Review by sydney-runkle — APPROVED** [review]
@alexdrydew,

Thanks for the contribution 🌟 ! Looks great.

**Pull request merged**